File: Generecidad/Ejercicio 2/Catalogo.java
import java.util.ArrayList;

public class Catalogo<T> {
    private ArrayList<T> elementos;
    
    public Catalogo() {
        elementos = new ArrayList<T>();
    }
    
    public void agregar(T elemento) {
        elementos.add(elemento);
    }
    
    public T buscar(int indice) {
        if (indice < 0 || indice >= elementos.size()) {
            System.out.println("no permitido");
            return null;
        }
        return elementos.get(indice);
    }
}

File: Generecidad/Ejercicio 2/Catalogo.py
from typing import Generic, TypeVar

T = TypeVar('T')

class Libro:
    def __init__(self, titulo: str, autor: str):
        self.titulo = titulo
        self.autor = autor
    
    def __str__(self) -> str:
        return f"{self.titulo} - {self.autor}"

class Producto:
    def __init__(self, nombre: str, precio: float):
        self.nombre = nombre
        self.precio = precio
    
    def __str__(self) -> str:
        return f"{self.nombre} - {self.precio} Bs"

class Catalogo(Generic[T]):
    def __init__(self):
        self.elementos = []
    
    def agregar(self, elemento: T) -> None:
        self.elementos.append(elemento)
    
    def buscar(self, indice: int) -> T:
        if indice < 0 or indice >= len(self.elementos):
            print("no permitido.")
            return None
        return self.elementos[indice]


libros = Catalogo[Libro]()
libros.agregar(Libro("Mañana llovera", "San DIego"))
libros.agregar(Libro("Vivir o Morir", "Dante Miguel"))
print(libros.buscar(0))
print(libros.buscar(1))
    
productos = Catalogo[Producto]()
productos.agregar(Producto("Tablet", 2100.0))
productos.agregar(Producto("Television", 5700.0))
print(productos.buscar(0))
print(productos.buscar(1))

